contribute/design/1.x/csi/20190805-csi-zfspv-volume-provisioning.md

```diff
@@ -109,48 +109,105 @@ spec:
 
 ### 2. CSI Node Publish
-When kubernetes schedules the application pod according to the Topology mentioned in the StorageClass, The CSI will get a NodePublish event. It will get all the volume properties from ZfsVolume custom resource and it will fire zfs create command and creates the volume. It will get the pool name from the ZfsVolume custom resource and create a volume on that pool with pvc name. Then it will format and mount the created zvol to the desired location and return.
+When the application pod is scheduled on a node, the CSI node-agent will get a NodePublish event. The driver will get the device path from the ZFSVolume custome resource and try to mount the file system as per user request given in the storage class. Once the ZFS volume dataset is created it will put a finalizer and return successful for the NodePublish event.
 
 - the kubernetes managed ZFS volume will look like this
 
 ```yaml
-apiVersion: v1
-items:
-- apiVersion: openebs.io/v1alpha1
-  kind: ZFSVolume
-  metadata:
-    creationTimestamp: "2019-08-25T08:02:14Z"
-    generation: 2
-    name: pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    namespace: openebs
-    resourceVersion: "13735"
-    selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/zfsvolumes/pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    uid: a691916c-c70e-11e9-8fa2-42010a80012d
-  spec:
-    blocksize: 4k
-    capacity: "4294967296"
-    compression: "on"
-    dedup: "on"
-    devicePath: /dev/zvol/zfspv-pool/pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-    ownerNodeID: gke-user-zfspv-default-pool-354050c7-wl8v
-    poolName: zfspv-pool
-    thinprovison: "yes"
-    volName: pvc-a6855135-c70e-11e9-8fa2-42010a80012d
-kind: List
+apiVersion: openebs.io/v1alpha1
+kind: ZFSVolume
 metadata:
-  resourceVersion: ""
-  selfLink: ""
+  creationTimestamp: "2019-11-03T06:08:02Z"
+  finalizers:
+  - zfs.openebs.io/finalizer
+  generation: 1
+  labels:
+    kubernetes.io/nodename: gke-pawan-zfspv-default-pool-8b577544-5wgl
+  name: pvc-4b083833-fe00-11e9-b162-42010a8001b1
+  namespace: openebs
+  resourceVersion: "147959"
+  selfLink: /apis/openebs.io/v1alpha1/namespaces/openebs/zfsvolumes/pvc-4b083833-fe00-11e9-b162-42010a8001b1
+  uid: 4b0fd8bc-fe00-11e9-b162-42010a8001b1
+spec:
+  blocksize: 4k
+  capacity: "4294967296"
+  compression: "on"
+  dedup: "on"
+  encryption: ""
+  keyformat: ""
+  keylocation: ""
+  ownerNodeID: gke-pawan-zfspv-default-pool-8b577544-5wgl
```

Comment:
If this value is changed will this volume get created on the updated node?

Reply:
no it wont, you need to remove the finalizer also. But why would someone change that because once volume is created on a node it should stay there only?